README.md

# rsMove
[![CRAN version](https://www.r-pkg.org/badges/version/rsMove)](https://CRAN.R-project.org/package=rsMove)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/last-month/rsMove?color=brightgreen)](https://CRAN.R-project.org/package=rsMove)
[![CRAN downloads](http://cranlogs.r-pkg.org/badges/grand-total/rsMove?color=brightgreen)](https://CRAN.R-project.org/package=rsMove)

### Why develop rsMove?

<p align="justify">
In the scope of movement ecology, Global Positioning Systems (GPS) have evolved significantly offering a unique insight into animal behavior. But understanding this behavior is dependent on our ability to comprehend the underlying environmental conditions that guides it. In this context, remote sensing becomes a fundamental tool. It provides information on the spatial and temporal variability of the landscape and provides us the means to understand the impact of environmental change over animal behavior. However, linking remote sensing and animal movement can be troublesome due to the differences in the spatial and temporal scales at which they are acquired (<a href="https://movementecologyjournal.biomedcentral.com/articles/10.1186/s40462-015-0036-7">Neuman et al, 2015</a>). As a consequence, methods that are sensitive to the constraints imposed by remote sensing in the analysis of animal movement are required. <i>rsMove</i> answers to this issue providing tools to query and analyze movement data using remote sensing.
</p>

<br>

### Installation
This gitHub is used as a basis for the improvement of *rsMove*. A stable release is available on <a href="https://cran.r-project.org/package=rsMove">CRAN</a> and can installed with:

```R
install.packages('rsMove')
```

<br>

### Examples
<p align="justify">
A paper on the package is currently under review (cross your fingers) and thus the vignette on the general applicability of rsMove was removed to avoid conflicts. Hopefully, I'll announce the paper here soon. For the time being, you can click <a href="https://cran.r-project.org/package=rsMove/vignettes/resource_suitability.html">here</a> for a vignette on the use of rsMove to predict environmental resource suitability following the methodology described in <a href="http://onlinelibrary.wiley.com/doi/10.1002/rse2.70/full">this</a> paper.
</p>

<br>

### Bug reports & contact

For bug reports, please use <a href="https://github.com/RRemelgado/rsMove/issues/"</a>. Feature requests and other contributions are also welcome.

<br>

### What else are we doing?
<p align="justify">
The Department of Remote Sensing of the University of Würzburg has developed other R packages that might interest you:
</p>

* <a href="https://bleutner.github.io/RStoolbox/">RStoolbox</a>
* <a href="https://github.com/cran/moveVis/">moveVis</a>
* <a href="https://github.com/RRemelgado/rsMove">fieldRS</a>
* <a href="https://github.com/RRemelgado/CAWaR">CAWaR</a>

<p align="justify">
Click <a href="http://remote-sensing.eu/">here</a> for news on our department.
</p>

<br>

### Aknowledgements
<p align="justify">
This initiative is part of the <a href="https://www.geographie.uni-wuerzburg.de/en/fernerkundung/research/completed-projects/opt4environment/">Opt4Environment</a> project and was funded by the German Aerospace Center (DLR) on behalf of the Federal Ministry for Economic Affairs and Energy (BMWi) with the research grant <b>50 EE 1403</b>. The movement data we used was provided by the Max Planck institute for Ornithology (MPIo). Click below to reach the involved parties.
</p>

<br>
<p align="justify">
<a href="https://www.geographie.uni-wuerzburg.de/en/fernerkundung/startseite/"><img width="150" height="100" src="https://www.uni-wuerzburg.de/typo3conf/ext/uw_sitepackage/Resources/Public/Images/uni-wuerzburg-logo.svg"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="https://www.dlr.de/eoc/en/desktopdefault.aspx/tabid-5278/8856_read-15911/"><img width="115" height="100" src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f5/DLR_Logo.svg/744px-DLR_Logo.svg.png"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="https://www.bmwi.de/Navigation/DE/Home/home.html"><img width="220" height="100" src="https://www.bmwi.de/SiteGlobals/BMWI/StyleBundles/Bilder/bmwi_logo_de.null?__blob=normal&v=9"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="https://www.orn.mpg.de/en/"><img width="200" height="100" src="https://www.molgen.mpg.de/188611/mpi_Seew_LogoText-1355515314.gif"></a>
</p>

<br>


NEWS.md
=================================================================================

#### rsMove 0.2.8

=================================================================================

### Changes
  * Altered projection string of example, spatial data due to proj6 migration

=================================================================================

#### rsMove 0.2.7

=================================================================================

### New
  * Added intime(), runmed2() and runmean2() c++ functions.
  * imgInt() allows for multi-band raster objects with observarion dates.
  * imgInt() can smooth the output using an user provided function.
  
### Changes:
  * imgInt() now depends on intime() and intime2()
  * ingInt() only accepts raster objects
  * segRaster() now depends on segMatrix()


=================================================================================

#### rsMove 0.2.6

=================================================================================

### Changes
  * sMoveRes() and tMoveRes() optimized.

### Fixes:
  * The size legend in plotMove() is now consistent.
  ' moveReduce() provides correct values when preserve.revisits was FALSE.

=================================================================================

#### rsMove 0.2.5

=================================================================================

### Changes
  * timeDir() returns a histogram ggplot object.
  * plausibilityTest() informs on the validation of each presence sample.
  * hotMove() returns a ggplot object with region polygons.
  * The output of hotMoveStats() was simplified.
  * moveReduce() allows the summary of movement data across temporal windows.

### Fixes:
  * checkOverlap() returned a warning when combining rasters and polygons.
  * Overall improvement of the detailed description of each function.
  * Simplification of variable names.

=================================================================================

#### rsMove 0.2.4

=================================================================================

#### New:
  * Added function checkOverlap() and plausibilityTest()
  
#### Changes:
  * Simplification of variable names in all functions

#### Fixes:
  * backSample(), labelSample() and imgInt() optimized
  * poly2sample now accepts RasterStack objects

=================================================================================

#### rsMove 0.2.3

=================================================================================

##### New:
  * No new functions added.
  
##### Changes:
  * Inclusion of plotting routines within functions.
  * Functions that plot results now provide an editable ggplot object.
  
##### Fixes:
  * Signficant improvements to the description of functions.
  * Updated function keywords to clarify their usage.

=================================================================================

#### rsMove 0.2.2

=================================================================================

##### New:
  * Introduced new functions
    * sMoveRes()
    * tMoveRes()
    * specVar()
    * getEnv()
    * proSat()
    * moveCloud()
    * satTime()
    * moveReduce()
    * plotMove()
 * imgInt(), moveSeg(), timeDir(), spaceDir(), linInt() now accept data frames as inputs.
 * hotMoveStats() now reports the temporal segment found within each region and their corresponding indices.
 * hotMoveStats() allows the analysis of temporal information without needing to run hotMove() first.
 * spaceDir() accepts categorical data.
 * spaceDir() and moveSeg() use an optional spatial buffer.
 * moveSeg(), timeDir(), spaceDir() and hotMoveStats() return plots.
 * dataQuery() requires an adjustable buffer specifying the size of search window before and after the target date.

##### Changes:
  * spaceDirSample() was renamed as spaceDir() to avoid confusion regarding its purpose.
  * timeDirSample() was renamed as timeDir() to avoid confusion regarding its purpose.
  * moveMovel() now requires data frames for presences and absences instead of shapefiles.
  * moveSeg(), spaceDir() and timeDir() no longer samples to single pixels internaly.
 
##### Fixes:
 * imgInt() allows that the interpolation is limited to the records in a point shapefile.
 * moseSeg() now recognizes categorical data.
 * BackSample() now assings background samples to the correct pixels.

=================================================================================

#### rsMove 0.1.0

=================================================================================

### Initial release to CRAN (2017-07-15) with the following functions:
  * backSample()
  * dataQuery()
  * hotMove()
  * hotMoveStats()
  * imgInt()
  * labelSample()
  * modelApply()
  * moveModel()
  * moveSeg()
  * poly2sample()
  * rsComposite()
  * sampleMove()
  * segRaster()
  * spaceDirSample()
  * timeDirSample()


R/plotMove.R
#' @title plotMove
#'
#' @description {Standardized plotting of environmental and temporal information for a set of coordinate pairs.}
#' @param x Vector of x coordinates.
#' @param y Vector of y coordinates.
#' @param size.var Optional. Controls the point size.
#' @param fill.var Optional. Controls the fill color.
#' @param var.type One of 'cont' or 'cat'. Defines the type of \emph{fill.var}.
#' @param var.names Character vector with names for \emph{size.var} and \emph{fill.var} to ve added to the plot.
#' @importFrom ggplot2 ggplot aes_string theme geom_bar scale_fill_gradientn xlab ylab theme_bw geom_point guides
#' scale_size_continuous scale_color_discrete scale_fill_gradientn scale_size_continuous guide_legend element_text element_blank
#' @importFrom grDevices colorRampPalette
#' @seealso \code{\link{dataQuery}} \code{\link{moveReduce}}
#' @return A \emph{ggplot} object.
#' @details {This function was designed to extent on other functions such as \code{\link{dataQuery}}, which provides environmental
#' information, and \code{\link{moveReduce}}, which provides information on e.g. the time spent per sample. Using these two functions
#' as an example, \emph{plotMove} can represent the relation between the elapsed time and the change in environmental conditions.}
#' @examples {
#'
#'  require(raster)
#'
#'  # read raster data
#'  r <- raster(system.file('extdata', '2013-07-16_ndvi.tif', package="rsMove"))
#'
#'  # read movement data
#'  data(shortMove)
#'
#'  # observation time
#'  time <- strptime(paste0(shortMove@data$date, ' ', shortMove@data$time), format="%Y/%m/%d %H:%M:%S")
#'
#'  # reduce amount of samples
#'  move.reduce <- moveReduce(shortMove, r, time)
#'
#'  # query data
#'  ov <- extract(r, move.reduce$points)
#'
#'  # plot output
#'  x <- move.reduce$points@data$x
#'  y <- move.reduce$points@data$y
#'  et <- move.reduce$points@data$elapsed.time
#'  op <- plotMove(x, y, size.var=et, fill.var=ov, var.type="cont")
#'
#' }
#' @export

#----------------------------------------------------------------------------------------------------------#

plotMove <- function(x, y, size.var=NULL, fill.var=NULL, var.type=NULL, var.names=NULL) {

  #----------------------------------------------------------------------------------------------------------#
  # 1. Check input data
  #----------------------------------------------------------------------------------------------------------#

  # check coordinates
  if (!is.numeric(x) | !is.numeric(y)) {stop('"x" and/or "y" not numeric')}
  if (length(x)!=length(y)) {stop('"x" and "y" have different lengths')}

  # check size variable
  if(!is.null(size.var)) {
    if (!is.numeric(size.var)) {stop('"size.var" is nof of a valid class')}
    if (length(size.var)!=length(x)) {stop('coordinates and "size.var" have different lengths')}}

  # check fill variable
  if (!is.null(fill.var)) {
    if (is.null(var.type)) {stop('"fill.var" is set. Please specify "var.type"')}
    if (!var.type%in%c('cont', 'cat')) {stop('"var.type" is not a recognized keyword')}
    if (length(fill.var)!=length(x)) {stop('coordinates and "fill.var" have different lengths')}}

  # check variable names
  if (is.null(var.names)) {
    if (var.type == 'cont') {var.names <- c('size', 'value')}
    if (var.type == 'cat') {var.names <- c('size', 'class')}
  } else {
    if (!is.character(var.names)) {stop('"var.names" is not a character object')}
    if (length(var.names) != 2) {stop('"var.names" should have two elements')}
  }

  # abort function if no variable is provided
  if (is.null(size.var) & is.null(fill.var)) {
    warning('neither "size.var" or "fill.var" were specfied. aborted.')
    return()
  }

  #----------------------------------------------------------------------------------------------------------#
  # 2. Determine variable limits/breaks
  #----------------------------------------------------------------------------------------------------------#

  # size breaks
  if (!is.null(size.var)) {tb <- seq(min(size.var), max(size.var), sd(size.var))}

  # color scheme for fill
  if (!is.null(fill.var)) {cr <- colorRampPalette(c("dodgerblue3", "khaki2", "forestgreen"))}

  #----------------------------------------------------------------------------------------------------------#
  # 3. Build plots
  #----------------------------------------------------------------------------------------------------------#

  # size and environmental data
  if (!is.null(size.var) & !is.null(fill.var)) {

    # build data frame
    df <- data.frame(x=x, y=y, size=size.var, fill=fill.var)

    # build plot
    if (var.type=="cont") {
      p <- ggplot(df) + theme_bw() + geom_point(aes_string(x="x", y="y", size="size", fill="fill"), color="black", pch=21) +
        guides(size=guide_legend(order=1, override.aes=list(shape=1)), col=guide_legend(order=0, override.aes=list(shape=15, size=6))) +
        theme(legend.text=element_text(size=10), panel.grid.major=element_blank(),
              panel.grid.minor=element_blank()) +
        scale_size_continuous(name=var.names[1], breaks=tb, range=c(2,12), labels=sprintf("%0.2f", tb)) +
        scale_fill_gradientn(name=var.names[2], colours=cr(10))}
    if (var.type=="cat") {
      df$fill <- factor(df$fill)
      p <- ggplot(df) + theme_bw() + geom_point(aes_string(x="x", y="y", size="size", color="fill")) +
        guides(size=guide_legend(order=1, override.aes=list(shape=1)), col=guide_legend(order=0, override.aes=list(shape=15, size=6))) +
        theme(legend.text=element_text(size=10), panel.grid.major=element_blank(),
              panel.grid.minor=element_blank()) +
        scale_size_continuous(name=var.names[1], breaks=tb, range=c(2,12), labels=sprintf("%0.2f", tb)) +
        scale_color_discrete(name=var.names[2])}

  }

  #----------------------------------------------------------------------------------------------------------#

  # only size
  if (!is.null(size.var) & is.null(fill.var)) {

    # build data frame
    df <- data.frame(x=x, y=y, size=size.var)

    # build plot
    p <- ggplot(df) + theme_bw() + geom_point(aes_string(x="x", y="y", size="size"), color="black", pch=21) +
      guides(fill=FALSE, col=guide_legend(override.aes = list(shape=15, size=6))) +
      theme(legend.text=element_text(size=10), panel.grid.major=element_blank(),
            panel.grid.minor=element_blank()) +
      scale_size_continuous(name=var.names[1], breaks=tb, range=c(2,12), labels=sprintf("%0.2f", tb))
  }

  #----------------------------------------------------------------------------------------------------------#

  # only environmental data
  if (is.null(size.var) & !is.null(fill.var)) {

    # build data frame
    df <- data.frame(x=x, y=y, fill=fill.var)

    # build plot
    if (var.type=="cont") {
      p <- ggplot(df) + theme_bw() + geom_point(aes_string(x="x", y="y", color="fill"), color="black", pch=21) +
        guides(col=guide_legend(override.aes = list(shape=15, size=6))) +
        theme(legend.text=element_text(size=10), panel.grid.major=element_blank(),
              panel.grid.minor=element_blank()) +
        scale_size_continuous(name=var.names[1], breaks=tb, range=c(2,12)) +
        scale_fill_gradientn(name=var.names[2], colours=cr(10))}
    if (var.type=="cat") {
      df$fill <- factor(df$fill)
      p <- ggplot(df, aes_string(x="x", y="y", color="fill")) + theme_bw() + geom_point() +
        guides(col=guide_legend(override.aes = list(shape=15, size=6))) +
        theme(legend.text=element_text(size=10), panel.grid.major=element_blank(),
              panel.grid.minor=element_blank()) +
        scale_size_continuous(name=var.names[1], breaks=tb, labels=sprintf("%0.2f", tb)) +
        scale_color_discrete(name=var.names[2])}


  }

  return(p)

}

